### PR TITLE
python3Packages.spacy_models: add updateScript

### LINKED ISSUE #186217: SpaCy models break with version 3.4.0
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: python310Packages.spacy_models (not found in nix search, but available in REPL)

current version: 3.3.0
desired version: 3.4.0

###### Notify maintainers

maintainers: @MoritzBoehme 

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by MoritzBoehme**
###### Description of changes

Adds an update script to update all spacy models and updates all models to 3.4.0 using the update script.
Closes #186217

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by MoritzBoehme** [comment]
@ofborg build python310Packages.spacy_models

File: pkgs/development/python-modules/spacy/default.nix [hunk 1]
```diff
@@ -31,6 +31,8 @@
 
 buildPythonPackage rec {
   pname = "spacy";
+  # when updating also run the update script for the models
+  # nix-shell maintainers/scripts/update.nix --argstr package python3Packages.spacy_models.en_core_web_sm
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spacy/default.nix:35]: Please add it to passthru.updateScript and remove this comment.

File: pkgs/development/python-modules/spacy/models.nix [hunk 2]
```diff
@@ -32,6 +34,24 @@ let
...
+
+      pushd pkgs/development/python-modules/spacy/ || exit
+
+      jq -r '.[] | .pname' models.json | while IFS= read -r pname; do
+        if [ "$(jq --arg pname "$pname" -r '.[] | select(.pname == $pname) | .version' models.json)" == "${spacy.version}" ]; then
+          continue
+        fi
+
+        newHash="$(nix-prefetch-url "https://github.com/explosion/spacy-models/releases/download/$pname-${spacy.version}/$pname-${spacy.version}.tar.gz")"
+        jq --arg newHash "$newHash" --arg pname "$pname" '[(.[] | select(.pname != $pname)), (.[] | select(.pname == $pname) | .sha256 = $newHash | .version = "${spacy.version}")] | sort_by(.pname)' models.json > tmp.json && mv tmp.json models.json
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spacy/models.nix:49]:
```suggestion
        jq --arg newHash "$newHash" --arg pname "$pname" '[(.[] | select(.pname != $pname)), (.[] | select(.pname == $pname) | .sha256 = $newHash | .version = "${spacy.version}")] | sort_by(.pname)' models.json | sponge models.json
```

File: pkgs/development/python-modules/spacy/models.nix [hunk 3]
```diff
@@ -32,6 +34,24 @@ let
 
     pythonImportsCheck = [ pname ];
 
+    passthru.updateScript = writeScript "update-spacy-models" ''
+      #!/usr/bin/env nix-shell
+      #!nix-shell -i bash -p jq
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spacy/models.nix:39]:
```suggestion
    #!${stdenv.shell}
    set -o errexit
    PATH=${lib.makeBinPath [ jq ]}
```

File: pkgs/development/python-modules/spacy/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/spacy/default.nix]: If a package has no update script r-ryantm tries to use some default methods. 

I think you need to add an updater for spacy which would probably be nix-update.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -10366,7 +10366,7 @@ in {
 
   spacy-loggers = callPackage ../development/python-modules/spacy-loggers { };
 
-  spacy_models = callPackage ../development/python-modules/spacy/models.nix { };
+  spacy_models = callPackage ../development/python-modules/spacy/models.nix { inherit (pkgs) jq; };
```
Comment (MoritzBoehme) [inline pkgs/top-level/python-packages.nix:10369]: Weirdly, when calling the updateScript of spacy_models `jq` got interpreted as `python3Packages.jq`. This is the best fix I could think of.

File: pkgs/development/python-modules/spacy/default.nix [hunk 5]
```diff
@@ -85,7 +90,19 @@ buildPythonPackage rec {
     "spacy"
   ];
 
-  passthru.tests.annotation = callPackage ./annotation-test { };
+  passthru = {
+    updateScript = writeScript "update-spacy" ''
+    #!${stdenv.shell}
+    set -o errexit
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spacy/default.nix:96]:
```suggestion
    set -eou pipefail
```

File: pkgs/development/python-modules/spacy/models.nix [hunk 6]
```diff
@@ -32,6 +38,28 @@ let
 
     pythonImportsCheck = [ pname ];
 
+    passthru.updateScript = writeScript "update-spacy-models" ''
+      #!${stdenv.shell}
+      set -o errexit
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/spacy/models.nix:43]:
```suggestion
      set -eou pipefail
```

**Pull request merged**